README.md
GNOME Telnet v2.4
--------------------------------------------------------------------------------------------------------------
This is an old GNOME telnet/rsh/ssh GUI. Last included in Redhat 7.1 and Mandriva 2007.0 for i586/x86_64

Copyright (C) 2000 the Free Software Foundation
Written by:
 [x] Kenth Andersson <d99kan@du.se> (v1.0)
 [x] Prana <*******@****.***> (v1.1, v1.2, v2.0, v2.1, v2.3, v2.4)

Language translations by:
 [x] Matthias Wimmer <matthias@matthias-wimmer.de> (Deutsch translation)
 [x] Takeshi Aihana <aihana@muc.biglobe.ne.jp> (Japanese translation)
 [x] Grega Bremec <gregab@gbsoft.org> (Slovene translation + helpful patches for v2.3)

RPM packages fixes & patches by:
 [x] Tim Powers <timp@redhat.com> (Redhat specs and patch to documentation location so that it works)
 [x] Vincent Danen <vdanen@mandrakesoft.com> (Mandrake specs)


Description
--------------------------------------------------------------------------------------------------------------
GNOME Telnet is a nice 3-in-1 frontend to the telnet, ssh, and rlogin client

Requirement
--------------------------------------------------------------------------------------------------------------
- autoconf >= v2.13
- automake >= v1.4
- gettext >= v0.10.34
- gtk+ >= v1.20
- gnome-core >= 1.0.55
- X Window 3.3.6
- Linux (of course) - other than Linux we can't give you a tech support

* Optional:
- SSH or OpenSSH (any version)

COMPILING & INSTALLING GTELNET (***** _IMPORTANT_ *****)
--------------------------------------------------------------------------------------------------------------

*** IMPORTANT NOTICE ***
Mandrake 7.02 doesn't have the right gettext package. Please get the lattest gettext package from
ftp://ftp.flash.net/pub/linux/distributions/mandrake/7.1/Mandrake/RPMS/gettext-0.10.35-13mdk.i586.rpm
Users running Mandrake 7.1 and 7.2 and Corporate Server shouldn't have any problems

You have to log in as a root user

** Easy way to compile it: run ./compile_install (you have to be logged in as ROOT)

Hard way compile it: 
1) "./autogen.sh --prefix=/usr"
   In case it doesn't compile, use "./configure --prefix=/usr" instead
2) make
3) strip src/gnome-telnet
4) make install (you have to log in as root)
5) ./install_shared (you have to log in as root)
6) If you want to reduce the size of the binary, type: "strip /usr/bin/gnome-telnet"
   But I wouldn't recommend doing that for debug purpose. However, this v2.0 is pretty stable enough ;)

The executable can be found in "/usr/bin/gnome-telnet"
I have tested this on Redhat Linux 6.2, Mandrake Linux 7.1 with HelixGnome v1.2
If you encounter that you can't compile this on Slackware or Debian, please e-mail me with the detail of the error messages

TODO LIST:
--------------------------------------------------------------------------------------------------------------
- This is almost the final release of Gnome-Telnet
- Check if SSH is installed or not
- I will add compatibility for OpenSSH for ssh-keygen

History (the same as ChangeLog file)
--------------------------------------------------------------------------------------------------------------
6th January 2000 (v2.4) - Prana
- This is only a maintenance release
- Fixed - it now correctly disconnects when you exits from your session
- Fixed - the color preferences
- Clean up code a little tiny bit

4th December 2000 (v2.3) - Prana
- Added command line argument
- Added Slovene translation by Grega Bremec
- Added (from users request) so that you can specify user@some.host.com and the SSH dialog won't show up
- SunOS patch
- Fixed for specifying the cyphertype (Grega Bremec <- very helpful person)
- Updated documentation
- Thanks to Grega Bremec <http://www.gbsoft.org> for the patches for SSH and the Slovene translation
- Thanks to the unknown person (I apologize - I lost your e-mail), but I remember your putenv() thing
  when I read the e-mail so that this gnome-telnet can work in SunOS too
- Thanks to Tim Powers for the fix for documentation location
- Fix Mandrake's version of the documentation location too - how come everybody is changing to /usr/share/doc ?
- Okay this is the really final one ...

22nd October 2000 (v2.2) - Prana
- IMPORTANT: Change in the binary executable from "Gnome_Telnet" to "gnome-telnet"
- Improved documentation
- Added SSH user login dialog - granted requests from numerous users :)
- Added terminal compatibility with aixterm, hpterm, iris-ansi, vt100, and more!
- Added more and more error checking
- Fixed a very tiny bug: color mapping problem that makes the color blurry when running Pine
- Optimized the non-redundant code generated by the IDE - thus the binary executable
  is now smaller than the previous release
- It seems that it's the final stage of the development, it is stable. It has been
  tested with OpenSSH 2.2.0p1, OpenSSH 1.1.2, SSH 2, telnet, and rlogin
- Moved the documentation to /usr/doc/gnome-telnet-2.2 instead of using /usr/share
- Thank you to people who have suggested features thus making this little program works better :)
- Now it's time to move on and start continuing the development of SubNetwork Explorer
- If you are so inclined, please help poor little kids
  by sponsoring them - visit http://www.worldvision.com :-)

30th August 2000 (v2.1) - Prana
- Improved compatibility with SSH2
- Added Deutsch (Germany) translation by Matthias Wimmer <matthias@matthias-wimmer.de>
- Added documentation and help file
- Added history combo box for the host name entry
- Fixed environment problem (now it's set to xterm)
- Fixed color mapping problem
- Fixed a bug in the preferences dialog
- Fixed Mandrake 7.1 installation bug. Now the RPM package will be seperated from Mandrake 7.02
- Fixed memory allocation problem
- Fixed scrollbar problem when resizing/scrolling the window

28th July, 2000 (v2.0) - Prana
- Added support for SSH
- Added support for Rlogin
- Added scrollbar
- Bug in input entry in the toolbar is fixed (ALL)
- Now if you're connected the status bar will display some kind of light
- And other numerous bug fixes and other several enhancements
- Major code cleanup - (it used to be so messy!)
- Thanks to Takeshi Aihana <aihana@muc.biglobe.ne.jp> for adding the Japanese locale

26th July, 2000 (v2.0) - Prana
-Start coding again

2nd June, 2000 (v1.2) - Prana
- Background images is now supported
- Bug in input entry in the preferences dialog is now fixed (ALL)
- And other numerous bug fixes and other several enhancements

31st May, 2000 (v1.1) - Prana
- Enhancement of GTelnet
- All user interface are changed
- Preferences Dialog (for telnet options)
- etc...

27th September 1999 (v1.0) - Kenth Andersson
- First initial public release of GTelnet
DOCUMENTATION
--------------------------------------------------------------------------------------------------------------
Click "Help" button :)




NEWS
May 31, 2000:
- Prana: Finished v1.1

2nd June, 2000:
- Prana: Finished v1.2 with numerous enhancements & bugfixes:-)

26th July, 2000:
- Prana: Started coding again

28th July, 2000:
- Prana: Finished v2.0

30th July, 2000:
- Prana: Finished v2.1

22nd October, 2000
- Prana: This is the final version: v2.1 Enjoy!

src/main.c
/*
 *	GTelnet
 *	by: 	Kenth Andersson <d99kan@du.se> (Initial development)
 *		Prana <*****@*****.***> (v1.1, v1.2, v2.0, v2.1, v2.2, v2.3, v2.4) 
 *
 *	Developed partially with Glade <http://glade.pn.org>
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#define ALL_CHKBOX
#define NO_XPM_PLEASE

#include <gnome.h>
#include <stdio.h>
#include <unistd.h>
#include <zvt/zvtterm.h>
#include <string.h>
#include <stdlib.h>
#include "interface.h"
#include "support.h"

static void read_configuration_file(void)
/* Read/Write Configuration */
{
  options = (char *) malloc(150);
  fontstr=NULL;
  the_interface_initialize();
  if (fontstr==NULL)
  {
     fontstr= (char *) malloc(100);
     fontstr=strcpy(fontstr,"-dec-terminal-medium-r-normal-*-*-140-*-*-c-*-iso8859-1");
  }
}

static void usage(const char *progname) {
/*
   Display the usage
*/
 char *logo = 
"\nGnome Telnet v2.3\n\
Visit http://***.*****.*** for more software!\n\n";
 
 char *usage= 
 "%s options [ssh_user]@host_name [port]\n\
 [options]:\n\
 \t-? : show this help screen\n\
 \t-h : show this help screen\n\
 \ttelnet: open a telnet connection to the targetted host\n\
 \tssh: secure shell connection to the host specified\n\
 \trlogin: remote login ala old BSD style\n\n\
 For example:\n\
 \t\"gnome-telnet ssh user@some_host.com\"\n\
 \t\"gnome-telnet telnet some_host.com\"\n";

 fprintf(stderr,logo);
 fprintf(stderr,usage,progname);
 exit(1);
}

int main (int argc, char *argv[])
/*
   Main program
*/
{
#ifdef ENABLE_NLS
  bindtextdomain (PACKAGE, PACKAGE_LOCALE_DIR);
  textdomain (PACKAGE);
#endif
  if (argc > 1) // if it's >= 2
  {
     if ( !strcmp(argv[1],"/?") || !strcmp(argv[1],"-?") || !strcmp(argv[1],"-h") || !strcmp(argv[1],"--help") ||
     argc < 3 || argc > 4)
        usage(argv[0]);
  }
  gnome_init ("gnome-telnet", VERSION, argc, argv);
  main_window = create_main_window (argc,argv[1],argv[2],argv[3]);
  read_configuration_file(); 
  gtk_widget_show (main_window);
  show_term ();
  gtk_main ();
  gnome_entry_save_history(GNOME_ENTRY(hostname_entry));
  // no mem leak
  free(options);
  free(fontstr);
  return 0;
}


src/interface.h
#ifndef __INTERFACE_H
#define __INTERFACE_H
/*
 *	GTelnet
 *	by: 	Kenth Andersson <d99kan@du.se> (Initial development v1.0)
 *		Prana <*****@*****.***> (v1.1, v1.2, v2.0, v2.1, v2.2, v2.3, v2.4)
 *
 *	Developed partially with Glade <http://glade.pn.org>
 */


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <netdb.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <gnome.h>
#include <zvt/zvtterm.h>
#include "stoerer.h"
#include "support.h"

#define vertbox 0
#define horzbox 1
#define max_telnet_desc 13
#define max_ssh_desc 15
#define max_rlogin_desc 8

  GtkWidget* create_main_window (int x_count,char *x_choice, char *x_hostname, char *x_port);
  GtkWidget* create_property_box (void);
  GtkWidget* create_message_box (gchar *s);
  GtkWidget* create_about_dialog (void);
  GtkWidget* create_ssh_dialog (void);

  void the_interface_initialize(void);
  void getoption(void);
  void show_term(void);

  // Handling events ...  
  void on_term_died (ZvtTerm *term);
  void on_connect_button_clicked              (GtkButton       *button, gpointer         user_data);
  void on_disconnect_button_clicked           (GtkButton       *button, gpointer         user_data);
  void on_preferences_button_clicked          (GtkButton       *button, gpointer         user_data);
  void on_about_button_clicked                (GtkButton       *button, gpointer         user_data);
  void on_pref_ok_button_clicked              (GtkButton       *button, gpointer         user_data);
  void on_pref_help_button_clicked            (GtkButton       *button, gpointer         user_data);
  void on_pref_apply_button_clicked           (GtkButton       *button, gpointer 	 user_data);
  void on_dock_leave 			      (GtkWidget       *button, gpointer 	 user_data);
  void on_about_dialog_cancel_clicked         (GtkButton       *button, gpointer         user_data);
  void on_ssh_login_ok_button_clicked         (GtkButton       *button, gpointer         user_data);
  void on_ssh_login_cancel_button_clicked     (GtkButton       *button, gpointer         user_data);
  
  // For handling with configuration files...
  void setarg(char *);
  void set_string_opt(char *flag, char *s);
  void set_direct_string_opt(char *flag, char *value);
  void set_flag_opt(char *flag, char *s);
  void setbutton(GtkWidget *button, gchar *s);
  void setbutton_string(GtkWidget *checkbox, GtkWidget *entry, char *s);
  void setaccordingto_string(GtkWidget *checkbox, GtkWidget *entry, gchar *s);
  void setaccordingto(GtkWidget *obj, gchar *s);

  
  int connection_type;
  int ssh_encrypt_type;
  gboolean bgimg_transparent;
  gboolean swap_del_backspace;
  GtkWidget *bgcolor_picker;
  GtkWidget *fgcolor_picker;
  GtkWidget *execolor_picker;
  GtkWidget *dircolor_picker;
  GtkWidget *fontpicker;
  GtkWidget *bgimage_fileentry;
  GtkWidget *transparent_checkbox;
  GtkWidget *bgimage_combo_entry;
  GtkWidget *edit_ssh_key_entry;

  GtkWidget *main_window;
  GtkWidget *about_dialog;
  GtkWidget *property_box;
  GtkWidget *ssh_keygen;  
  GtkWidget *term;
  GtkWidget *term2;

  GdkColor	gdk_color_forground;
  GdkColor	gdk_color_background;
  GdkColor	gdk_color_green;
  GdkColor	gdk_color_blue;

  GtkWidget *connect_via;
  GtkWidget *main_window;
  GtkWidget *hostname_entry;
  GtkWidget *port_entry;
  GtkWidget *tmp_toolbar_icon;
  GtkWidget *app_statusbar;


/* Telnet options */
  GtkWidget *tos_entry;  
  GtkWidget *initchar_entry;  
  GtkWidget *atype_entry;
  GtkWidget *realm_entry;

  GtkWidget *atype_entry;
  GtkWidget *tos_entry; 
  GtkWidget *realm_entry;
  GtkWidget *initchar_entry;
  GtkWidget *tracefiles_entry; // the gnome file entry
  GtkWidget *tracefile_entry; // the input
  GtkWidget *telnet_option;

/* SSH Options */
  GtkWidget *ssh_encryption_option;
  GtkWidget *ssh_encryption_option_menu;  
  GtkWidget *ssh_esc_char_entry;
  GtkWidget *rsa_identity_entry;
  GtkWidget *rsa_identity;
  GtkWidget *ssh_debug_level;
  GtkWidget *swap_del_key_checkbox;

/* RLogin */
  GtkWidget *rlogin_ticket_entry;
  GtkWidget *init_rlogin_esc_entry;
  GtkWidget *hostname_combo_entry;
  // v2.2
  GtkWidget *term_type_entry;
  GtkWidget *term_type_combo_entry;
  GtkWidget *ssh_login_entry;
  GtkWidget *ssh_login_combo_entry;
  GtkWidget *ssh_dialog;
  
  char *options; 		/* options for telnet */
  char *term_environ;		/* term environment variable */
  char *fontstr; 		/* Font name for the term -- Kenth -- unchanged*/
  char *image_file_location; 	/* Background image for Telnet */
  char *terminal_type;		/* Set the terminal type */
  char **pargv;
  
  int run_option; // if used within cmd-line

  gchar *host; /* hostname */
  gchar *port; /* port number */
  gchar *ssh_user; /* SSH user */
  char *arguments[16]; /* argument */
  int n_counter;
 
  struct opt_widget
  {
   GtkWidget *widget;
   char *name;
   gchar *desc;
  };

 struct terminal_colors_type
 {
     gushort red;
     gushort green;
     gushort blue;
 };

#endif


po/cat-id-tbl.c
/* Automatically generated by po2tbl.sed from gnome-telnet.pot.  */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include "libgettext.h"

const struct _msg_ent _msg_tbl[] = {
  {"", 1},
  {"8-bit data path (both output and input)", 2},
  {"8-bit data path on output", 3},
  {"Stop any character from being recognized as an escape character", 4},
  {"\
Forward a forwardable copy of the local credentials to the remote system (if \
Kerberos V5 is used)", 5},
  {"Attempt automatic login", 6},
  {"Specify a user interface similar to rlogin", 7},
  {"Turn on encryption of the data stream", 8},
  {"Set the initial value of the debug flag to TRUE", 9},
  {"Set the initial telnet escape character to:", 10},
  {"Set the IP type-of-service  (TOS)  option  to the value:", 11},
  {"Disable the type of authentication: ", 12},
  {"Request  Kerberos tickets for the remote host in the realm:", 13},
  {"Trace file :", 14},
  {"Redirect input from /dev/null", 15},
  {"Disable authentication agent forwarding", 16},
  {"Disable X11 connection forwarding", 17},
  {"Allocate a tty even if command is given", 18},
  {"Display verbose debugging messages", 19},
  {"Don't allocate a privileged port", 20},
  {"Don't display any warning messages", 21},
  {"Fork into background after authentication", 22},
  {"Enable compression", 23},
  {"Allow remote hosts to connect to local forwarded ports", 24},
  {"Don't request a session", 25},
  {"Set SSH2 debug level to:", 26},
  {"Identity file for public key authentication to:", 27},
  {"Set escape character to:", 28},
  {"\
Select encryption algorithm (For OpenSSH: only '3des' and 'blowfish' are \
supported)", 29},
  {"Allows an eight-bit input data path at all times", 30},
  {"Stops any character from being recognized as an escape character", 31},
  {"Turns off all Kerberos authentication", 32},
  {"Allows the rlogin session to be run in \"litout\" mode", 33},
  {"Turns on socket debugging", 34},
  {"Turns on DES encryption for all data passed via the rlogin session", 35},
  {"Request Kerberos tickets for the remote host in the realm:", 36},
  {"Background Color", 37},
  {"Foreground Color", 38},
  {"Executable Files Color", 39},
  {"Directories Color", 40},
  {"Display Font", 41},
  {"Background Image", 42},
  {"GTelnet", 43},
  {"Select one of the options", 44},
  {" Host name ", 45},
  {" Port ", 46},
  {"port number -- leave blank for default", 47},
  {"Connect", 48},
  {"Connect to the remote host", 49},
  {"Disconnect", 50},
  {"Disconnect from the remote host", 51},
  {"Preferences", 52},
  {"Change your settings", 53},
  {"Help", 54},
  {"Documentation", 55},
  {"Information", 56},
  {"About GTelnet", 57},
  {"Click here for a complete documentation", 58},
  {"\
GTelnet is a frontend for the following programs:\n\
\n\
- Telnet:  to communicate with another host using the TELNET protocol. \n\
\n\
- SSH:  ssh (Secure Shell) is a program for logging into a remote machine \
and for executing commands on a remote machine.  It is intended to replace \
rlogin and rsh, and provide secure encrypted communications between two \
untrusted hosts over an insecure network.  X11 connections and arbitrary \
TCP/IP ports can also be forwarded over the secure channel. (Copied from SSH \
manual)\n\
\n\
- RLogin: rlogin is a program that is a little bit similar to telnet. It is \
intended to make remote administration over local area network easier, but \
it can be very insecure. It is recommended that you use SSH instead of \
rlogin.", 59},
  {"Copyright (C) 2000 the Free Software Foundation", 60},
  {"GTelnet is a frontend to the ole good telnet, ssh, and rlogin", 61},
  {"Authors: ", 62},
  {"Authors", 63},
  {"\
Today, there's a lot of poor children out there who need your help.\n\
\n\
For a dollar a day, you can help change a child's life forever. With your \
sponsorship, you will help provide basic needs like clean water, schools, \
clinics and agricultural training that will lift your sponsored child, \
his/her family, and community - out of poverty.\n\
\n\
If you have the resources available and if you're willing to help those who \
are less fortunate, please sponsor a child.", 64},
  {"For more information, please visit:", 65},
  {"http://www.WorldVision.com", 66},
  {"Please Help Needy Children", 67},
  {"Trace file", 68},
  {"Telnet Options", 69},
  {"SSH Options", 70},
  {"RLogin Options", 71},
  {"Terminal type:", 72},
  {"Display Settings", 73},
  {"SSH Login", 74},
  {"Username:", 75},
  {"\
If you don't want to see this login box, type user@hostname in the host name \
entry, and to restore the previous settings you can click the pulldown \
button next to it to retrieve the history.", 76},
  {"Please fill in the host name :-)", 77},
  {"Not connected", 78},
  {"Couldn't find pixmap file: %s", 79},
  {"Couldn't create pixmap from file: %s", 80},
};

int _msg_tbl_length = 80;
